File: src/dp/_322_coin_change.rs
struct Solution;

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let mut dp = vec![amount + 1; (amount + 1) as usize];
        dp[0] = 0;

        for i in 0..=amount {
            for c in &coins {
                if i - c < 0 {
                    continue;
                }
                dp[i as usize] = std::cmp::min(dp[i as usize], 1 + dp[(i - *c) as usize]);
            }
        }
        if dp[amount as usize] == amount + 1 {
            return -1;
        }
        dp[amount as usize]
    }
}

File: hot-100/322_coinChange.go
package hot_100

import "math"

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	dp[0] = 0
	for i := 1; i < amount+1; i++ {
		dp[i] = math.MaxInt
	}

	for i := 0; i < len(coins); i++ {
		for j := coins[i]; j < amount+1; j++ {
			if dp[j-coins[i]] != math.MaxInt {
				dp[j] = min(dp[j], dp[j-coins[i]]+1)
			}
		}
	}

	if dp[amount] == math.MaxInt {
		return -1
	}
	return dp[amount]
}
